### PR TITLE
normalize iam policy actions

### PR DESCRIPTION
**Pull request by smcavallo**
Signed-off-by: smcavallo <smcavallo@hotmail.com>

### Description of your changes
Enhanced policy comparison.  See linked issue.
Converts `Action` with a single item from array to string for comparison purposes:
`"Action":["sts:AssumeRoleWithWebIdentity"]` to `"Action":"sts:AssumeRoleWithWebIdentity"` 

In effect this is the same permission but the current comparison interpret them as different.

Fixes #
https://github.com/crossplane/provider-aws/issues/932

I have:

- [X] Read and followed Crossplane's [contribution process].
- [X] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
* Unit tests

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Review by chlunde — COMMENTED** [review]
Thanks for looking at this!

I think the linked issue is for IAM Role assume policy conditions, while this change is for IAM policies. This highlights another issue, there are IAM policies everywhere, not just in the IAM API group but for many APIs there are resource based policies too.

I think it would be nice to solve the issue in the shared helper, ideally without regular expressions. Do you think it is possible to fix the shared helper and change iam policies to use that instead? Perhaps as an option to `cmp.Diff` (POC: https://play.golang.org/p/uZqVdENJcIZ - but this applies the transformation for too many JSON keys, like Condtions, which I'm not sure about. Maybe FilterPath can ensure we only transform in safe places.)

https://github.com/crossplane/provider-aws/blob/master/pkg/clients/aws.go#L739-L768

**Comment by smcavallo** [comment]
@chlunde - I 100% agree.  Will propose:
- `IAMRole` and `IAMPolicy` be processed through the same function.
- Conditions/Actions/Resources - single value arrays be converted to strings for comparisons
- Ignore sort/ordering be implemented as well
- This can be achieved through custom Marshal/Unmarshal instead of regex - will solve the sort/ordering and can be applied to only the fields where both ways are supported (ex. `Actions|Resources|etc`) instead of all

I did not take that approach since it is a much larger change and I did not know how Marshaling would affect policies/roles in the wild or if there were edge cases which would break if implemented.  There aren't enough unit tests around testing here, and one positive about the change is many more tests can be added.

File: pkg/clients/iam/iampolicy.go [hunk 1]
```diff
@@ -45,14 +47,44 @@ func IsPolicyUpToDate(in v1alpha1.IAMPolicyParameters, policy iamtypes.PolicyVer
...
 	}
 
 	return cmp.Equal(compactIAMPolicy, compactSpecPolicy), nil
 }
+
+// replaceActionArray converts Actions with a single item from an array to a string
+func replaceActionArray(compactJSON string) string {
+	// Ex.  Convert "Action": ["sts:AssumeRole"] -> "Action": "sts:AssumeRole"
+	// But ignore "Action": ["sts:AssumeRole", "sts:GetFederationToken" ] since there are multiple actions
+	r := regexp.MustCompile("\"Action\":\\[(\\S*?)\\]")
```
Comment (MisterMX) [inline pkg/clients/iam/iampolicy.go:79]: I don't think a regex is suited here. The parser from #1772 could be used here as well.

**Comment by MisterMX** [comment]
Closing in favor of #1772.
